### PR TITLE
[WIP] Doxygen Docs for Nixpkgs Internals

### PR DESCRIPTION
**Pull request by lukeadams**
###### Motivation for this change
This PR demonstrates a low friction way to autogenerate documentation for Nixpkgs lib functions.
Doxygen does a pretty decent job picking up the correct function name (although it thinks they are variables) under the C++ generator.

[A demo can be seen here.](http://people.tamu.edu/~luke.adams/doxygen-nixpkgs/files.html)

I haven't converted all of strings.nix, but the functions with code-formatted examples have been converted.
screen shot 2018-02-03 at 9 50 08 am

#### TODO
- [ ] Examples
  - [ ] Example header
  - [ ] [Syntax highlighting in examples](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdcode)
- [ ] Investigate alternate docstring generators
  - [ ] Doxygen (javadoc)
  - [ ] Sphinx (reStructuredText)
    * [Add new languages](https://samprocter.com/2014/06/documenting-a-language-using-a-custom-sphinx-domain-and-pygments-lexer/)
        * [Domaintools](https://bitbucket.org/birkenfeld/sphinx-contrib/src/e758073384efd1ed5ed1e6286301b7bef71b27cf/domaintools/?at=default)
  - [ ] [haddock](https://haskell-haddock.readthedocs.io/en/latest/markup.html) 
- [ ] Integrate docbook output into Nixpkgs manual
- [ ] Document internal functions
  - [ ] lib
  - [ ] stdenv
  - [ ] shell scripts

  

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by FRidh** [comment]
Very nice! I have a personal preference for Sphinx, but regardless I think this is very good to have and we should get this in. I wonder though what others think of changing the format of the docstrings.

**Comment by grahamc** [comment]
(I  added the python tag to check a hypothesis w.r.t. the incorrect label added by ofborg. I'll remove it shortly, sorry for the noise)

**Comment by vcunat** [comment]
Sphinx takes data via doxygen, at least in projects I know about.  I had thought about this long ago, but I didn't realize it could work to use a different language :heart:

**Comment by lukeadams** [comment]
@Fridh I totally agree! Sphinx definitely has a nicer looking theme by default. Doxygen lets you create a custom template, so I was thinking of mimicking the NixOS.org styling. I've got the header on the demo page [here](http://people.tamu.edu/~luke.adams/doxygen-nixpkgs/strings_8nix.html).

@vcunat Sort of a happy accident that it works since Doxygen is in c++ mode. Other modes may work better, although the "best" would be to write a Nix language definition. Glad you like it!

File: lib/strings.nix [hunk 1]
```diff
@@ -1,183 +1,263 @@
-/* String manipulation functions. */
+/** \file strings.nix
+ * String manipulation functions.
+ */
 { lib }:
 let
 
 inherit (builtins) length;
-
+/** @private */
```
Comment (lukeadams) [inline lib/strings.nix]: notice how I had to mark this as private. since doxygen is in c++ mode it thinks these are variables.

**Comment by FRidh** [comment]
> Sphinx takes data via doxygen, at least in projects I know about.

Right. That's still an option then.

@lukeadams could you look how to integrate this in the Nixpkgs manual? Doxygen supports docbook output. http://doxygen.nl/manual/config.html#cfg_generate_docbook

When we have that available, we could consider looking further than just `lib`, like also the `trivial-builders` and shell scripts like `makeWrapper`.

**Comment by lukeadams** [comment]
@FRidh With that option enabled, I just have to add Doxygen generation to doc/default.nix and tell the docstring builder to look at its output.

I've now labelled strings.nix as a member of `lib` so it shows up [here](http://people.tamu.edu/~luke.adams/doxygen-nixpkgs/modules.html). All functions in other files will show up there as long as they are marked as members of lib. I'm not sure how to get Doxygen to generate docs from .sh files, but it would be trivial to create a markdown file for each one and document it that way

**Comment by lukeadams** [comment]
(Sorta) got it linked into the manual [here](http://people.tamu.edu/~luke.adams/nixpkgs-manual/manual.html#idp140699425324856).

Needs some massaging to make it look good though.

**Comment by lexleogryfon** [comment]
I've been looking for such documentation and happy that someone finally commits to writing it. 
All Nixpkgs builtin functions should have clear API documentation reference like standard libraries in other languages, specifying type of arguments, type of return, usage example. Syntax highlight would be nice. 
Will appreciate to see project alive and completed, even if it will not be merged yet to master.

**Comment by oxij** [comment]
I like the general idea, but I find the markup to be uber ugly.

We already have the documentation that many people (including me) hate to edit because of docbook XML, now it seems to me that you want make me hate to edit the lib because of this Javadoc-style @trash.

No offense, my problem is not with this PR specifically, but with @trash in general.

I would love to have exactly this idea implemented with `haddock` instead of `doxygen`. `haddock` docstrings are beautiful.
Also, `haddock` parses Haskell, nix is more like Haskell than like C++.

**Comment by lukeadams** [comment]
@FRidh Honestly I think Sphinx would be the best option with a custom domain (link up top) which would add Nix support to it. Doxygen requires too much hacking to work reliably and its Docbook output requires quite a bit of massaging to embed in the manual.

I'll probably start with the [C++ domain](https://github.com/sphinx-doc/sphinx/blob/2a65ffeef5c107c19084fabdd706cdff3f52d93c/sphinx/domains/cpp.py) and modify it to support Nix.
Thoughts?

**Comment by lukeadams** [comment]
@FRidh I'll have some time this weekend for this PR.

Any preference on comment style ( \@ vs \\ )?
```c++
/**
* @brief Your Brief Desc
*/

/**
* \brief Your Brief Desc
*/
```
---
Apparently Doxygen has options (`HIDE_UNDOC_MEMBERS` and `HIDE_UNDOC_CLASSES`) to only show functions which have valid Docstring comments, removing the need for `\@cond ... \@endcond`, which fixes the main issue I had with it.

**Comment by lukeadams** [comment]
I'm going to close this as I no longer have the time to work on it. 
May pursue it in the future by using the Nix Lex/Yacc parser to write a frontend for Sphinx, or maybe create a standalone tool "NixDoc" in the same vein as rustdoc

**Pull request closed without merging**